### ISSUE TITLE
learning problem with gradParams:zero()

### ISSUE THREAD

**Issue by osilvam** [torch/torch7#993]
Hi,

First, sorry for my english. I have some problems with the zero function of gradParams obtained through getParameters function. The model in the following code (optim training example https://github.com/torch/optim/blob/master/doc/intro.md) don't learn anything if i call gradParams:zero() inside the for statment (as indicated in the example), but the model does learn if i call gradParams:zero()  before the for statment. So, the example is wrong ? or i don't understand what happening

Thanks in advance

require 'nn'

model = nn.Sequential()  -- make a multi-layer perceptron
inputs = 2; outputs = 1; HUs = 20 -- parameters
model:add(nn.Linear(inputs, HUs))
model:add(nn.Tanh())
model:add(nn.Linear(HUs, outputs))

criterion = nn.MSECriterion()

batchSize = 128
batchInputs = torch.DoubleTensor(batchSize, inputs) -- or CudaTensor for GPU training
batchLabels = torch.DoubleTensor(batchSize)         -- or CudaTensor for GPU training

for i = 1, batchSize do
	local input = torch.randn(2)     -- normally distributed example in 2d
	local label
	if input[1] * input[2] > 0 then  -- calculate label for XOR function
		label = -1
	else
		label = 1
	end
	batchInputs[i]:copy(input)
	batchLabels[i] = label
end

params, gradParams = model:getParameters()

local optimState = {learningRate = 0.01}

require 'optim'

for epoch = 1, 250 do
	function feval(params)
		gradParams:zero() 

		local outputs = model:forward(batchInputs)
		local loss = criterion:forward(outputs, batchLabels)
		local dloss_doutputs = criterion:backward(outputs, batchLabels)		
		model:backward(batchInputs, dloss_doutputs)

		return loss, gradParams
	end

	optim.sgd(feval, params, optimState)
end

x = torch.Tensor(2)
x[1] =  0.5; x[2] =  0.5; print(model:forward(x))
x[1] =  0.5; x[2] = -0.5; print(model:forward(x))
x[1] = -0.5; x[2] =  0.5; print(model:forward(x))
x[1] = -0.5; x[2] = -0.5; print(model:forward(x))
